### PR TITLE
[Example] NGNN for ogbl

### PR DESCRIPTION
**Pull request by Ereboas**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes

- [x] Add an example for running NGNN on ogbl-ddi, ogbl-collab and ogbl-ppa.

## Note
Since the structures of NGNN layers of 3 datasets(ogbl-ddi, ogbl-collab and ogbl-ppa) are different from each other(In this PR there're Linear+Relu once or twice after a single GNN layer; but there're Linear+Relu+Linear+sigmoid structure for some ogbn experiments and more structures in the original paper), and it is not that suitable to adjust the structures by command scripts, so I do not merge 3 codes to 1 though they look similar.

### CODE REVIEW

**Comment by mufeili** [comment]
How different are the implementations across different datasets? If most of the code is shared, then perhaps we can merge them into one folder just like PR #4291

**Comment by Ereboas** [comment]
> How different are the implementations across different datasets? If most of the code is shared, then perhaps we can merge them into one folder just like PR #4291

Different in dataset preprocessing and ngnn structure settings. And I've tried to merge the similar parts in the latest commit.

File: examples/pytorch/ogb/ngnn/README.md [hunk 1]
```diff
@@ -0,0 +1,387 @@
+# GraphSage/GCN (FULL)
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/README.md:1]: NGNN + GraphSage/GCN

You don' need to have "(FULL)" for now.

**Comment by mufeili** [comment]
Have you tried re-running the scripts for some time to ensure that the code is not broken after the merge?

File: examples/pytorch/ogb/ngnn/README.md [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+
+## Introduction
+
+This is an example of implementing [NGNN](https://arxiv.org/abs/2111.11638) for link prediction in DGL.
+
+We use a model-agnostic methodology, namely Network In Graph Neural Network (NGNN), which allows arbitrary GNN models to increase their model capacity.
+
+The script in this folder experiments full-batch GCN/GraphSage (with/without NGNN) on the dataset: ogbl-ddi, ogbl-collab and ogbl-ppa.
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/README.md:9]: I think it suffices to include the code/numbers related to with NGNN.

Comment (mufeili) [inline examples/pytorch/ogb/ngnn/README.md:9]: "dataset" -> "datasets"

File: examples/pytorch/ogb/ngnn/README.md [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+#### Reproduction of performance
+
+- GCN + NGNN
+```{.bash}
+python main.py --dataset ogbl-ddi --device 0 --ngnn_type input --epochs 800 --dropout 0.5 --num_layers 2 --lr 0.0025 --batch_size 16384 --runs 50
+```
+
+- GraphSage + NGNN
+```{.bash}
+python main.py --dataset ogbl-ddi --device 1 --ngnn_type input --use_sage --epochs 600 --dropout 0.25 --num_layers 2 --lr 0.0012 --batch_size 32768 --runs 50
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/README.md:140]: Do we still need scripts then?

File: examples/pytorch/ogb/ngnn/README.md [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+      <td>48.22% ± 7.00%</td>
+      <td>82.56% ± 4.03%</td>
+      <td>89.48% ± 1.68%</td>
+      <td>65.95% ± 1.16%</td>
+      <td>70.24% ± 0.50%</td>
+      <td>72.54% ± 0.62%</td>
+      <td rowspan=2>1,487,361</td>
+   </tr>
+   <tr>
+      <td>GCN+NGNN(ours; 50runs)</td>
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/README.md:82]: It suffices to have the numbers for "(ours; 50 runs)" and perhaps "(paper)" if you want. For the rest numbers, people should refer to the paper.

**Comment by mufeili** [comment]
The numbers you got for `ogbl-ddi` and `ogbl-ppa` are significantly better. Could you briefly explain what adjustments make a significant contribution to the increase?

File: examples/pytorch/ogb/ngnn/main.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+
+    # dataset setting
+    parser.add_argument('--dataset', type=str, default='ogbl-ddi', choices=['ogbl-ddi', 'ogbl-collab', 'ogbl-ppa'])
+
+    # device setting
+    parser.add_argument('--device', type=int, default=0, help='GPU device ID. Use -1 for CPU training.')
+
+    # model structure settings
+    parser.add_argument('--use_sage', action='store_true', help='If not set, use GCN by default.')
+    parser.add_argument('--ngnn_type', type=str, default="none", choices=['none', 'input', 'hidden', 'output', 'all'], help="You can set this value from 'none', 'input', 'hidden' or 'all' to apply NGNN to different GNN layers.")
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/main.py:328]: It seems that we only need to support two options -- 'input' and 'hidden'

File: examples/pytorch/ogb/ngnn/main.py [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+    parser.add_argument('--num_layers', type=int, default=3, help='number of GNN layers')
+    parser.add_argument('--hidden_channels', type=int, default=256)
+    parser.add_argument('--dropout', type=float, default=0.0)
+    parser.add_argument('--batch_size', type=int, default=64 * 1024)
+    parser.add_argument('--lr', type=float, default=0.001)
+    parser.add_argument('--epochs', type=int, default=400)
+
+    # training settings
+    parser.add_argument('--use_valedges_as_input', action='store_true', help='Use training + validation edges for inference on test set.')
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/main.py:337]: It seems that you do not use this argument in `README.md`. Then you can remove it and the related code.

File: examples/pytorch/ogb/ngnn/main.py [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+    parser.add_argument('--log_steps', type=int, default=1)
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/main.py:338]: Ditto

File: examples/pytorch/ogb/ngnn/main.py [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+    parser.add_argument('--eval_steps', type=int, default=1)
+    parser.add_argument('--runs', type=int, default=10)
+    global args; args = parser.parse_args()
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/main.py:341]: It's not very common to use "global args". Can we drop it?

File: examples/pytorch/ogb/ngnn/main.py [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+    print(args)
+
+    device = f'cuda:{args.device}' if args.device != -1 and torch.cuda.is_available() else 'cpu'
+    device = torch.device(device)
+
+    dataset = DglLinkPropPredDataset(name=args.dataset)
+    data = dataset[0]
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/main.py:348]: Perhaps you can rename it to `g` to be more consistent with other DGL examples.

File: examples/pytorch/ogb/ngnn/main.py [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+        self.fc_adj2 = Linear(hidden_channels, out_channels)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        self.conv.reset_parameters()
+        gain = torch.nn.init.calculate_gain('relu')
+        torch.nn.init.xavier_uniform_(self.fc_adj.weight, gain=gain)
+        torch.nn.init.xavier_uniform_(self.fc_adj2.weight, gain=gain)
+        for bias in [self.fc_adj.bias, self.fc_adj2.bias]:
+            if bias is not None:
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/main.py:71]: bias is never None. Also perhaps move `import math` to the top.

File: examples/pytorch/ogb/ngnn/main.py [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+                import math
+                stdv = 1.0 / math.sqrt(bias.size(0))
+                bias.data.uniform_(-stdv, stdv)
+
+    def forward(self, g, x):
+        x = self.conv(g, x)
+
+        global args
+        if args.dataset != 'ogbl-ddi':
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/main.py:80]: I'll prefer passing the dataset name in constructing the model object.

File: examples/pytorch/ogb/ngnn/main.py [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+            r = best_result[:, 3]
+            print(f"   Final Test: {r.mean():.2f} ± {r.std():.2f}")
+
+
+class NGNN_GCNConv(torch.nn.Module):
+    def __init__(self, in_channels, hidden_channels, out_channels):
+        super(NGNN_GCNConv, self).__init__()
+        self.conv = GraphConv(in_channels, hidden_channels)
+        self.fc_adj = Linear(hidden_channels, hidden_channels)
+        self.fc_adj2 = Linear(hidden_channels, out_channels)
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/main.py:62]: The use of `adj` is a bit confusing, perhaps just call them `self.fc` and `self.fc2`.

File: examples/pytorch/ogb/ngnn/main.py [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        self.conv.reset_parameters()
+        gain = torch.nn.init.calculate_gain('relu')
+        torch.nn.init.xavier_uniform_(self.fc_adj.weight, gain=gain)
+        torch.nn.init.xavier_uniform_(self.fc_adj2.weight, gain=gain)
+        for bias in [self.fc_adj.bias, self.fc_adj2.bias]:
+            if bias is not None:
+                import math
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/main.py:143]: Ditto for L142-143

File: examples/pytorch/ogb/ngnn/main.py [hunk 14]
```diff
@@ -0,0 +1,437 @@
...
+                stdv = 1.0 / math.sqrt(bias.size(0))
+                bias.data.uniform_(-stdv, stdv)
+
+    def forward(self, g, x):
+        x = self.conv(g, x)
+
+        global args
+        if args.dataset != 'ogbl-ddi':
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/main.py:151]: Ditto

File: examples/pytorch/ogb/ngnn/main.py [hunk 15]
```diff
@@ -0,0 +1,437 @@
...
+        return x
+
+
+class NGNN_SAGEConv(torch.nn.Module):
+    def __init__(self, in_channels, hidden_channels, out_channels,
+                 *, reduce):
+        super(NGNN_SAGEConv, self).__init__()
+        self.conv = SAGEConv(in_channels, hidden_channels, reduce)
+        self.fc_adj = Linear(hidden_channels, hidden_channels)
+        self.fc_adj2 = Linear(hidden_channels, out_channels)
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/main.py:133]: Ditto for L132-133

File: examples/pytorch/ogb/ngnn/main.py [hunk 16]
```diff
@@ -0,0 +1,437 @@
...
+
+    evaluator = Evaluator(name=args.dataset)
+    loggers = {
+        'Hits@20': Logger(args.runs, args),
+        'Hits@50': Logger(args.runs, args),
+        'Hits@100': Logger(args.runs, args),
+    }
+
+    for run in range(args.runs):
+        torch.cuda.empty_cache()
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/main.py:394]: Do we have an issue without `torch.cuda.empty_cache()`?

**Comment by mufeili** [comment]
I've done a first pass. Overall a great job.

File: examples/pytorch/ogb/ngnn/README.md (hunk 3 above)
Comment (Ereboas) [reply examples/pytorch/ogb/ngnn/README.md:140]: I think so. It follows the rules of ogb (https://ogb.stanford.edu/docs/leader_overview/  --> How Leaderboards Work? -->code --> Please add README and provide enough instruction **(i.e., exact command)** to reproduce the submitted result.)

File: examples/pytorch/ogb/ngnn/main.py (hunk 16 above)
Comment (Ereboas) [reply examples/pytorch/ogb/ngnn/main.py:394]: No, it's just a habit of mine. I will remove it in the next commit.

File: examples/pytorch/ogb/ngnn/main.py (hunk 6 above)
Comment (Ereboas) [reply examples/pytorch/ogb/ngnn/main.py:337]: OGB allows to use validation edges (together with train edges) as input only for the ogbl-collab dataset. I think it may be useful to improve the performance. Anyway, I will remove it in the next commit.

**Comment by Ereboas** [comment]
> Have you tried re-running the scripts for some time to ensure that the code is not broken after the merge?

Yes. All scripts are tested.

> The numbers you got for `ogbl-ddi` and `ogbl-ppa` are significantly better. Could you briefly explain what adjustments make a significant contribution to the increase?

Done.

I have updated a new commit, please review it at your convenience. Thanks.

**Comment by mufeili** [comment]
Could you also make a review? @rudongyu

File: examples/pytorch/ogb/ngnn/README.md (hunk 3 above)
Comment (mufeili) [reply examples/pytorch/ogb/ngnn/README.md:102]: Yes, but the command with `python main.py ...` seems to be a duplicate of the `.sh` files?

**Review by mufeili — APPROVED** [review]
I'm good.

File: examples/pytorch/ogb/ngnn/main.py [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+        for bias in [self.fc.bias, self.fc2.bias]:
+            stdv = 1.0 / math.sqrt(bias.size(0))
+            bias.data.uniform_(-stdv, stdv)
+
+    def forward(self, g, x):
+        x = self.conv(g, x)
+
+        if self.dataset != 'ogbl-ddi':
+            x = F.relu(x)
+            x = self.fc(x)
```
Comment (BarclayII) [inline examples/pytorch/ogb/ngnn/main.py:81]: Can we have some other flag like `self.num_layers==2` instead of `self.dataset`?  `self.dataset` in a NN module looks ugly...

**Review by BarclayII — COMMENTED** [review]
My feeling is that we are blending the code for multiple datasets and multiple architecture variants together in the same script again, just like what we did before the huge refactor.  Should we split them to multiple files, one for each dataset?  @mufeili

**Comment by mufeili** [comment]
> My feeling is that we are blending the code for multiple datasets and multiple architecture variants together in the same script again, just like what we did before the huge refactor. Should we split them to multiple files, one for each dataset? @mufeili

Which huge refactor do you refer to? I feel that the OGB submissions tend to be more complex than normal examples.

**Comment by BarclayII** [comment]
@mufeili Like in `graphsage` examples we used to handle different datasets in the same script.  I personally would favor having one script for one dataset (like in `ogbn-products` directory).  I also saw SIGN that handles multiple datasets at once though so I guess there's no standard for how to organize OGB examples.  So I'm not objecting the current way.

**Comment by mufeili** [comment]
> @mufeili Like in `graphsage` examples we used to handle different datasets in the same script. I personally would favor having one script for one dataset (like in `ogbn-products` directory). I also saw SIGN that handles multiple datasets at once though so I guess there's no standard for how to organize OGB examples. So I'm not objecting the current way.

I see. I guess we might will revisit it in the future. Thanks.

**Comment by mufeili** [comment]
@Ereboas 

Two minor issues left:

1. "Yes, but the command with python main.py ... seems to be a duplicate of the .sh files?"
2. The comment by @BarclayII regarding if it is possible to replace `self.dataset != 'ogbl-ddi'` with a check based on `self.num_layers`.

File: examples/pytorch/ogb/ngnn/README.md (hunk 3 above)
Comment (Ereboas) [reply examples/pytorch/ogb/ngnn/README.md:102]: Is it a good solution to remove the ```scripts``` folder(and its contents) and related description in ```README.md```?

Comment (mufeili) [reply examples/pytorch/ogb/ngnn/README.md:102]: Yes, I think so.

**Comment by Ereboas** [comment]
I've tested ngnn codes in the latest commit(d7a1230).

File: examples/pytorch/ogb/ngnn/README.md [hunk 18]
```diff
@@ -17,7 +17,7 @@ dgl>=0.8
 
 ## Usage
 
-run the scripts in `./scripts` folder, e.g.
+run the command-line scripts, e.g.
 ```{.bash}
 scripts/run_ddi.sh
 ```
```
Comment (mufeili) [inline examples/pytorch/ogb/ngnn/README.md:23]: I think you can remove this section. The scripts are gone.
